File: project/sql/jimmer-sql-test/jimmer-sql-test-model/src/main/java/org/babyfish/jimmer/sql/transaction/ServiceB.java
package org.babyfish.jimmer.sql.transaction;

import java.util.concurrent.CompletionException;
import org.babyfish.jimmer.sql.JSqlClient;

@TargetAnnotation(Component.class)
public class ServiceB {

    protected final JSqlClient sqlClient;

    public ServiceB(JSqlClient sqlClient) {
        this.sqlClient = sqlClient;
    }

    @Tx
    public void a() {}

    @Tx(Propagation.REQUIRES_NEW)
    public void b() {}

    @Tx
    public void c() throws CompletionException {}
}

File: project/sql/jimmer-sql-test/jimmer-sql-test-model-kotlin/src/main/kotlin/org/babyfish/jimmer/sql/kt/transaction/ServiceB.kt
package org.babyfish.jimmer.sql.kt.transaction

import java.util.concurrent.CompletionException
import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.transaction.Propagation
import org.babyfish.jimmer.sql.transaction.Tx

open class ServiceB(
    protected val sqlClient: KSqlClient
) {
    @Tx
    open fun a() {}

    @Tx(Propagation.REQUIRES_NEW)
    protected open fun b() {}

    @Throws(CompletionException::class)
    @Tx
    open fun c() {}
}
